### ISSUE TITLE
Some reports are "randomly" missing

### ISSUE THREAD

**Issue by ldez** [mgechev/revive#1306]
**Describe the bug**

Some reports are "randomly" missing

**To Reproduce**

```bash
git clone -q git@github.com:prometheus/prometheus.git
cd prometheus
git checkout 701d13abf
```

```console
$ revive ./... | wc -l
723
$ revive ./... | wc -l
723
$ revive ./... | wc -l
723
$ revive ./... | wc -l
723
$ revive ./... | wc -l
720
$ revive ./... | wc -l
723
$ revive ./... | wc -l
723
$ revive ./... | wc -l
723
$ revive ./... | wc -l
723
$ revive ./... | wc -l
723
$ revive ./... | wc -l
723
$ revive ./... | wc -l
720
$ revive ./... | wc -l
723
...
```

As you can see, sometimes there are 723 reports and sometimes there are only 720 reports.

**Expected behavior**

Report consistency.

**Additional context**

For now, I didn't find the root cause but I think the problem is related to a concurrency issue.
I will try to find the root cause.

**Comment by ldez** [comment]
The run (not the compilation) is at least 4 times slower.

```
 revive ./... | wc -l
==================
WARNING: DATA RACE
Write at 0x00c000e14020 by goroutine 231:
  github.com/mgechev/revive/lint.(*Package).TypeCheck()
      /home/ldez/sources/golangci-lint-linters/revive/lint/package.go:132 +0x833
  github.com/mgechev/revive/rule.(*TimeNamingRule).Apply()
      /home/ldez/sources/golangci-lint-linters/revive/rule/time_naming.go:25 +0x164
  github.com/mgechev/revive/lint.(*File).lint()
      /home/ldez/sources/golangci-lint-linters/revive/lint/file.go:109 +0x286
  github.com/mgechev/revive/lint.(*Package).lint.func1()
      /home/ldez/sources/golangci-lint-linters/revive/lint/package.go:196 +0x10b
  golang.org/x/sync/errgroup.(*Group).Go.func1()
      /home/ldez/sources/go/pkg/mod/golang.org/x/sync@v0.12.0/errgroup/errgroup.go:78 +0x91

Previous read at 0x00c000e14020 by goroutine 230:
  github.com/mgechev/revive/lint.(*Package).TypeOf()
      /home/ldez/sources/golangci-lint-linters/revive/lint/package.go:153 +0x75a
  github.com/mgechev/revive/rule.(*UnexportedReturnRule).Apply()
      /home/ldez/sources/golangci-lint-linters/revive/rule/unexported_return.go:44 +0x6e9
  github.com/mgechev/revive/lint.(*File).lint()
      /home/ldez/sources/golangci-lint-linters/revive/lint/file.go:109 +0x286
  github.com/mgechev/revive/lint.(*Package).lint.func1()
      /home/ldez/sources/golangci-lint-linters/revive/lint/package.go:196 +0x10b
  golang.org/x/sync/errgroup.(*Group).Go.func1()
      /home/ldez/sources/go/pkg/mod/golang.org/x/sync@v0.12.0/errgroup/errgroup.go:78 +0x91

Goroutine 231 (running) created at:
  golang.org/x/sync/errgroup.(*Group).Go()
      /home/ldez/sources/go/pkg/mod/golang.org/x/sync@v0.12.0/errgroup/errgroup.go:75 +0x124
  github.com/mgechev/revive/lint.(*Package).lint()
      /home/ldez/sources/golangci-lint-linters/revive/lint/package.go:195 +0x11a
  github.com/mgechev/revive/lint.(*Linter).lintPackage()
      /home/ldez/sources/golangci-lint-linters/revive/lint/linter.go:156 +0x557
  github.com/mgechev/revive/lint.(*Linter).Lint.func1()
      /home/ldez/sources/golangci-lint-linters/revive/lint/linter.go:107 +0x24b
  golang.org/x/sync/errgroup.(*Group).Go.func1()
      /home/ldez/sources/go/pkg/mod/golang.org/x/sync@v0.12.0/errgroup/errgroup.go:78 +0x91

Goroutine 230 (running) created at:
  golang.org/x/sync/errgroup.(*Group).Go()
      /home/ldez/sources/go/pkg/mod/golang.org/x/sync@v0.12.0/errgroup/errgroup.go:75 +0x124
  github.com/mgechev/revive/lint.(*Package).lint()
      /home/ldez/sources/golangci-lint-linters/revive/lint/package.go:195 +0x11a
  github.com/mgechev/revive/lint.(*Linter).lintPackage()
      /home/ldez/sources/golangci-lint-linters/revive/lint/linter.go:156 +0x557
  github.com/mgechev/revive/lint.(*Linter).Lint.func1()
      /home/ldez/sources/golangci-lint-linters/revive/lint/linter.go:107 +0x24b
  golang.org/x/sync/errgroup.(*Group).Go.func1()
      /home/ldez/sources/go/pkg/mod/golang.org/x/sync@v0.12.0/errgroup/errgroup.go:78 +0x91
==================
Found 1 data race(s)
720
```

**Comment by ldez** [comment]
🤔 I think I know how to fix that.

**Comment by denisvmedia** [comment]
> The run (not the compilation) is at least 4 times slower.

Yeah, that's expected, it's not recommended for production anyway. Does it still produce race condition issues even with your change?

**Comment by ldez** [comment]
I found the problem 🎉
